README.rst

```diff
@@ -0,0 +1,760 @@
+Context
+=======
+
+We build and release software by massively consuming and producing software
+packages such as NPMs, RPMs, Rubygems, etc.
+
+Each package manager, platform, type or ecosystem has its own conventions and
+protocols to identify, locate and provision software packages.
+
+
+Problem
+=======
+
+When tools, APIs and databases process or store multiple package types, it is
+difficult to reference the same software package across tools in a uniform way.
+
+For example, these tools, specifications and API use relatively similar
+approaches to identify and locate software packages, each with subtle
+differences in syntax, naming and conventions:
+
+- Grafeas uses a scheme, namespace, name and version in a URL-like string
+- Here.com OSRK uses a package manager, name and version field and a colon-
+  separated URL-like string
+- JFrog XRay uses a scheme, namespace, name and version in a URL-like string
+- Libraries.io uses a platform, name and version
+- OpenShift fabric8 analytics uses ecosystem, name and version
+- ScanCode and AboutCode.org use a type, name and version
+- SPDX has an appendix for external repository references and uses a type and a
+  locator with a type-specific syntax for component separators in a URL-like
+  string
+- versioneye uses a type, name and version
+
+
+Solution
+========
+
+A `purl` or package URL is an attempt to standardize existing approaches to
+reliably identify and locate software packages.
+
+A `purl` is a URL string used to identify and locate a software package in a
+mostly universal and uniform way across programing languages, package managers,
+packaging conventions, tools, APIs and databases.
+
+Such a package URL is useful to reliably reference the same software package
+using a simple and expressive syntax and conventions based on familiar URLs.
+
+
+purl
+~~~~~
+
+`purl` stands for **package URL**.
+
+A `purl` is a URL composed of six components::
+
+    type:namespace/name@version?qualifiers#subpath
```

Comment:
Is host or repository ubiquitous enough that it should be part of the URL rather than a qualifier? Some possible examples:

```
npm://registry.npmjs.org/Microsoft/redie@0.1.0
github://github.com/Microsoft/redie
github://myghe.server.example/Foo/bar
```

Without this there's an assumption of what the default registry is for the PURL, that could be a good thing (because it's resilient to change) but the downsides are that:
* The consumer would need to know what the default repository is
* The consumer needs to check the qualifier to determine if the repository has been overridden and fallback if necessary

Reply:
@iamwillbar I think the default repositories for a `type` are ubiquitous enough IMHO.
The idea is to encapsulate in a type/scheme all the many details and specifics of the general "protocol" of a package: for instance when I say `nuget` or `maven` this at once refers to a public repository, conventions and rules to create and build packages, the API, network protocols and more to talk to the repository, the format of the package manifest and the package archives, the clients that can talk to the repository and query, fetch, install, a dependency resolution mechanism, and so on: all this is quite major.

In contrast I tend to think of the repository host as secondary: there is a main public one, but there can be eventually an infinity of alternative repositories beyond the main one: private, local, mirrors, etc. And assuming that content-wise the package files or artifacts are verified as the same, this is the same package: therefore alternate repositories are minor. Not unimportant but minor and less important than a type, name and version IMHO. And as a lesser data bit, they should therefore come later in the `purl` rather than at the beginning in the `Authority` components of a URL/URI. 

As for the defaults, I tried to list some for the some known package types. We could very make this a data file that would be usable by anyone (actually it likely make sense also as there are some type-specific purl normalization rules, such as for pypi as reported by @jayfk )

Reply:
Think of it this way: package management CLI do not ask you to specify the default repository at all. They just use it by default unless you specific alternative hosts.

Reply:
@iamwillbar does my reply make sense to you?

Reply:
It does, but I think my fundamental concern is that you can have the same package identity in two different repositories but be different packages (this is commonly used as an attack vector in fact). By being explicit about the authority you state the canonical trusted location for that package, yes, you might find the package in other places but you can't assume they're the same without some other validation (i.e. comparing purls without the identity is not a safe operation). Does that make sense?

Reply:
@iamwillbar it makes perfect sense. 

You wrote:
> It does, but I think my fundamental concern is that you can have the same package identity in two different repositories but be different packages (this is commonly used as an attack vector in fact).

That's interesting.... would you have some pointers to examples of this? I would be quite interested to read more about it.

> By being explicit about the authority you state the canonical trusted location for that package, yes, you might find the package in other places but you can't assume they're the same without some other validation (i.e. comparing purls without the identity is not a safe operation). Does that make sense?

IMHO a `purl` should be dealing with with package files integrity and content verification: this is a complex topic that is `type`-specific and depends on many things already implemented in most if not all package managers with such thing as checksum or signature verification (including repositories public keys as is common for RPMs and Debian), ensuring the transport is TLS or SSH secured, etc. This is best managed outside of a `purl`as it is today. Yet there is provision to use `qualifiers` for this too. 

In the vast majority of the cases, comparing two`purl` strings should be fine. And if you need more, then again I think is best managed as an attribute outside of a purl ....

For instance, take Pypi: there can be several downloads available for a given name/version for various package archive formats and eventually native builds for diifferent os/arch/ as there https://pypi.python.org/pypi/lxml/

To get accurate down to the package archive level, you would want to get the download URL which looks this way:

https://pypi.python.org/packages/34/95/b3d5d1227561887eb691b7c6b0bc9aa7eaa752b13262df2d2f04c8633242/lxml-4.1.1-cp33-cp33m-macosx_10_6_x86_64.macosx_10_9_intel.macosx_10_9_x86_64.macosx_10_10_intel.macosx_10_10_x86_64.whl#md5=7a0fd79e4899eea1d32fe251e10105f3

And beside being a long and ugly name (here mostly to deal with some macOS ABI peculiarities) it contains two verifications: a trailing, weak md5 in the fragment and a string blake2b leading 512 bits checksum in the path

e.g. This and checksums are otherwise best managed outside IMHO

Reply:
I've to agree that for at least .debs and .rpms, I think the assumption that there's some kind of flat namespace does not hold. The tuple name_version_arch is only ever unique (possibly) within a specific archive. Which is what stuff like apt would expect, but some vendors might not even honor.

For .deb packages, there's also the concept of the target distribution and the one built in. So the Debian vim source package targets the Debian unstable distribution, and in Debian it's built there. But when Ubuntu takes that source package (which targeted Debian unstable, as can be seen from its debian/changelog) it gets rebuilt on the current Ubuntu distribution at the time, so the resulting binaries are different.

I think there's a distinction between many of the other types, because they include a canonical archive. While deb and rpm are just package formats.

For example in the dpkg world we've got the concept of a vendor (https://manpages.debian.org/dpkg-vendor), and actively encourage derivatives, forks and alternative implementations to declare their own vendor so that it can be referred to (https://wiki.debian.org/Derivatives/CensusFull). My plan is still to start embedding that vendor information into each .deb built so that we can distinguish them, and uniquely identify these different tuples (https://www.hadrons.org/~guillem/debian/docs/origin.proposal).

Also, even using something like "jessie" is not unique enough, and it's a moving attribute, as packages move from distribution to distribution (say uploaded to unstable, and migrated to testing, which points to stretch). And  while declaring the complete repository URL is certainly very cumbersome, and I'd say even an implementation detail that should not be leaking, I'd recommend instead including the vendor in the purl. So I think I'd qualify it by the vendor as part of the archive location, and then add possibly an attribute for the distribution currently in, which would be certainly optional. So the binary dpkg packages from Debian currently in stretch and Ubuntu could look like this?

  deb:debian/dpkg@1.19.0.4?arch=amd64&distro=stretch
  deb:ubuntu/dpkg@1.19.0.4?arch=amd64

Reply:
@guillemj Thank you ++ to chime in. As the lead maintainer for `dpkg`,  this is quite an honor! I owe you so much and I am so grateful for your code. 
Now, this makes 100% sense and I just pushed an update with this and you examples.

Reply:
@guillemj Actually, would both of these be built from the same exact source code? e.g.  would Ubuntu ever rebuild a package changing the sources and not change the version? because if this the exact same source, then may be the vendor could be a `qualifier` instead since then the "provenance" of the package would resolve to the exact same copyright and license, same bugs, same vulnerabilities, same everything source wise and may not be much different from an `arch`? 
(and hopefully when https://reproducible-builds.org is everywhere this may even be the same binaries?) 

So, may be this instead?

 
    deb:dpkg@1.19.0.4?arch=amd64&distro=stretch&vendor=debian
    deb:dpkg@1.19.0.4?arch=amd64&&vendor=ubuntu

Reply:
re:Ubuntu, right, I'd expect them not to reuse the name_version tuple for the source package for content that is distinct, and I'm not sure how much mechanism they have that would prevent that, even if due to accidents or omissions. But the reverse could happen. Someone packages something first in Ubuntu, and then someone else in Debian packages the same and does not notice, and reuses the same tuple. This actually applies to any Debian derivative, for which there might be even less tight relationships. So, I'd not even assume the uniqueness for source packages, even though between Debian and Ubuntu that should be mostly ok in principle. Also nothing says that the upstream tarball has not been modified when preparing the source package, to say, strip non-free stuff, which obviously depends on what every derivative considers non-free. :)

re:Reproducible-builds. That still cannot give guarantees beyond the same archive. Reproducible builds are defined as the same output for the same specified input, not for any random input. So, that's why we record a subset of the build environment. As long as the packages are not exactly the same, because for example Ubuntu has not synced all of them from Debian unstable, or even the builds happened on different days (so had different build-dependency versions installed) then the resulting binaries will most probably not be the same.

I guess in the end whether the vendor might need to be a fundamental part of the purl or just a qualifier, depends on the purpose of the purl. If it's to refer to an abstract reference of a package with certain attributes, then that might be fine. If it's to reference a specific instance of that package, that needs to be installed on a system, then probably the vendor is fundamental. I think there's widespread agreement within Debian that mixing and matching packages from different repos is unwise, more so when the different repos contain rebuilds of the same packages.

Reply:
@guillemj Thank you for the details. .... I am still weighing the pros and cons of using the Debian vendor as a namespace vs a qualifier key/value. Right now it is using the `namespace` based on your feedback, but I feel we would be better served by a qualifier: this still keeps distinct `purl`s for each vendor but also makes two `deb` packages with the same name and version sort close by which I kinda like as in the vast majority of the cases they would be distinct but eventually minor variants of the same code. In all cases, Debian would always sort before Ubuntu because of the alphabet (and the historical precedence as it should ;) )

Reply:
@guillemj thx. We are using vendor as namespace for deb and rpm now, thanks to your feedback!

Reply:
I saw `npm://registry.npmjs.org/Microsoft/redie@0.1.0` and immediately went: well, that's a bad idea.

As has been suggested elsewhere in the thread, eliding the registry when referring to the global registry solves the proxy problem. (One other thing, package owner should never be in an npm purl.)

So, something like:  `npm:redie@0.1.0`

Or on a private registry: `npm:redie@0.1.0?registry=private.corp.com`

The namespace thing conveniently is compatible w/ npm scopes. They do canonically start with an `@` which would be important to keep in mind when writing parsers. Thatd be like: `npm:@iarna/cli@1.0.0`